### PR TITLE
Full working example of 2D array with operator overloading.

### PR DESCRIPTION
**Pull request by quickfur**
Add fully working code example of a 2D array that overloads the indexing, slicing, and opDollar operators, since it may not be obvious what is the correct way to implement `opSlice` and `opDollar`.

### CODE REVIEW

**Comment by quickfur** [comment]
ping @MartinNowak

File: operatoroverloading.dd [hunk 1]
```diff
@@ -703,6 +703,98 @@ $(H2 $(LEGACY_LNAME2 ArrayOps, array-ops, Array Indexing and Slicing Operators O
...
+	sections following.)
+
+-------
+struct Array2D(E)
+{
+    E[] impl;
+    int stride;
+    int width, height;
+
+    this(int _width, int _height, E[] initialData = [])
```
Comment (MartinNowak) [inline operatoroverloading.dd]: Underscores in front of the fields are somewhat normal, but not in front of arguments.
Maybe use `this.width` instead.

File: operatoroverloading.dd [hunk 2]
```diff
@@ -703,6 +703,98 @@ $(H2 $(LEGACY_LNAME2 ArrayOps, array-ops, Array Indexing and Slicing Operators O
...
+
+    // Support for `x..y` notation in slicing operator.
+    int[2] opSlice(size_t dim)(int start, int end)
+        if (dim >= 0 && dim < 2)
+    {
+        return [start, end];
+    }
+
+    // Support `$` in slicing notation, e.g., arr[1..$, 0..$-1].
+    @property int opDollar(size_t dim)() if (dim==0) { return width; }
```
Comment (MartinNowak) [inline operatoroverloading.dd]: How about `opDollar(size_T dim:0)()`?

File: operatoroverloading.dd [hunk 3]
```diff
@@ -703,6 +703,98 @@ $(H2 $(LEGACY_LNAME2 ArrayOps, array-ops, Array Indexing and Slicing Operators O
...
+        result.stride = this.stride;
+        result.width = r1[1] - r1[0];
+        result.height = r2[1] - r2[0];
+
+        return result;
+    }
+    auto opIndex(int[2] r1, int j) { return opIndex(r1, [j, j+1]); }
+    auto opIndex(int i, int[2] r2) { return opIndex([i, i+1], r2); }
+
+    // Support for `x..y` notation in slicing operator.
```
Comment (MartinNowak) [inline operatoroverloading.dd]: Should mention "for the given dimension" or so.

File: operatoroverloading.dd [hunk 4]
```diff
@@ -703,6 +703,98 @@ $(H2 $(LEGACY_LNAME2 ArrayOps, array-ops, Array Indexing and Slicing Operators O
...
+    int[2] opSlice(size_t dim)(int start, int end)
+        if (dim >= 0 && dim < 2)
+    {
+        return [start, end];
```
Comment (MartinNowak) [inline operatoroverloading.dd]: Add `assert(start > 0 && end <= bounds[dim]);`

**Comment by quickfur** [comment]
Fixed all issues raised and rebased. Please review. Thanks!

**Comment by DmitryOlshansky** [comment]
Looks cute

**Comment by JackStouffer** [comment]
LGTM always nice to have a practical examples in docs :)

File: operatoroverloading.dd [hunk 5]
```diff
@@ -703,6 +703,100 @@ $(H2 $(LEGACY_LNAME2 ArrayOps, array-ops, Array Indexing and Slicing Operators O
...
+        if (dim >= 0 && dim < 2)
+    in { assert(start >= 0 && end <= this.opDollar!dim); }
+    body
+    {
+        return [start, end];
+    }
+
+    // Support `$` in slicing notation, e.g., arr[1..$, 0..$-1].
+    @property int opDollar(size_t dim : 0)() { return width; }
+    @property int opDollar(size_t dim : 1)() { return height; }
```
Comment (CyberShadow) [inline operatoroverloading.dd]: Didn't know you could do this with value parameters.

**Pull request merged**